sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/ProxyAuthenticationHandler.java

```diff
@@ -0,0 +1,287 @@
+        InetSocketAddress raddr = this.destinationAddress();
+        String hostString = HttpUtil.formatHostnameForHttp(raddr);
+        int port = raddr.getPort();
+        String url = hostString + ":" + port;
+        String hostHeader = (port != 80 && port != 443) ? url : hostString;
+        FullHttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.CONNECT, url,
+            Unpooled.EMPTY_BUFFER, false);
+
+        request.headers().set(HttpHeaderNames.HOST, hostHeader);
+
+        // Proxy is expected to have authentication requirements, attempt to authenticate.
+        if (challengeHandler != null) {
+            String authorizationHeader = createAuthorizationHeader();
+
+            if (!CoreUtils.isNullOrEmpty(authorizationHeader)) {
+                request.headers().set(PROXY_AUTHORIZATION, authorizationHeader);
+                ctx.channel().attr(PROXY_AUTHORIZATION_KEY).set(authorizationHeader);
+            }
+        }
+
+        return request;
+    }
+
+    /*
+     * Attempts to create a Proxy-Authorization header based on the AuthorizationChallengeHandler and ChallengeHolder
+     * configurations.
+     */
+    private String createAuthorizationHeader() {
+        /*
+         * Attempt to pipeline the request.
+         *
+         * Pipelining is only possible if another request has been authenticated using the
+         * AuthenticationChallengeHandler, it attempts to leverage state from the previous authentication request.
+         * This may fail and result in the server requesting authentication by returning a proxy authentication
+         * challenge.
+         */
+        String authorizationHeader = challengeHandler.attemptToPipelineAuthorization(PROXY_METHOD, PROXY_URI_PATH,
+            PROXY_ENTITY_BODY);
+
+        if (!CoreUtils.isNullOrEmpty(authorizationHeader)) {
+            return authorizationHeader;
+        }
+
+        ChallengeHolder proxyChallengeHolder = proxyChallengeHolderReference.get();
+
+        /*
+         * Check if a proxy authentication challenge has been passed to the channel. This occurs when a proxy
+         * authentication challenge response is seen by the connection observer and passed back to a future channel
+         * created from the same client.
+         */
+        if (proxyChallengeHolder != null) {
+            // Attempt to apply digest challenges, these are preferred over basic authorization.
+            List<Map<String, String>> digestChallenges = proxyChallengeHolder.getDigestChallenges();
+            if (!CoreUtils.isNullOrEmpty(digestChallenges)) {
+                authorizationHeader = challengeHandler.handleDigest(PROXY_METHOD, PROXY_URI_PATH, digestChallenges,
+                    PROXY_ENTITY_BODY);
+            }
+
+            // If digest challenges exist or all failed attempt to use basic authorization.
+            if (CoreUtils.isNullOrEmpty(authorizationHeader) && proxyChallengeHolder.hasBasicChallenge()) {
```

Comment:
Why do we fallback on basic auth if digest auth failed?

Reply:
The scenarios where basic is used are the following:

- All digest authentication challenges failed and a basic authentication challenge was returned.
- No digest authentication challenges were returned but a basic challenge was.

The overall idea is that the more secure authentication challenges are attempted first followed by less secure challenges if those fail. If you think we should ignore a basic challenge if digest fails I'd be on-board with that.

Reply:
Yeah, I would err on the side of caution before falling back to basic. If users request for this behavior, we can consider relaxing this but if the user started with `digest proxy` in mind, we should fail if that's not possible.

Reply:
With the current design there is no need to indicate Basic or Digest authentication, the `HttpClient` will use whatever challenge (if any) the server returns and it will prefer using Digest before Basic. The only configuration a user will determine is whether `ProxyOptions` has a username and password configured, if the configuration is there an `AuthorizationChallengeHandler` will be added into the `HttpClient` to handle Basic or Digest challenges and if there is no configuration and a challenge is received it will never be able to authorize itself. This is why there is a Basic fallback if all the Digest challenges are malformed or contain expected algorithms the JVM doesn't support.

This could be a place where the design is changed to allow configuring which type of authentication should be used and that could be where a user determines if Basic and Digest or just one of the two is supported for authorization.

Reply:
I am okay with keeping it simple and have a fallback on basic if digest fails. After all, the proxy server is the one that is going to send the challenge and it makes users life easier by not having to split hairs to decide whether to use digest or basic.